rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRACDIExtension.java

```diff
@@ -0,0 +1,249 @@
+
+    public void observe(@Observes AfterBeanDiscovery event, BeanManager beanManager) throws IOException, ClassNotFoundException {
+        Index index = classPathIndexer.createIndex();
+
+        List<AnnotationInstance> annotations = index.getAnnotations(DotName.createSimple("javax.ws.rs.Path"));
+
+        for (AnnotationInstance annotation : annotations) {
+            ClassInfo classInfo;
+            AnnotationTarget target = annotation.target();
+
+            if (target.kind().equals(AnnotationTarget.Kind.CLASS)) {
+                classInfo = target.asClass();
+            } else if (target.kind().equals(AnnotationTarget.Kind.METHOD)) {
+                classInfo = target.asMethod().declaringClass();
+            } else {
+                continue;
+            }
+
+            Class<?> clazz = getClass().getClassLoader().loadClass(classInfo.name().toString());
+
+            LRAParticipant participant = processJavaClass(clazz);
+            if (participant != null) {
+                participants.put(participant.getId().getName(), participant);
+                Set<Bean<?>> participantBeans = beanManager.getBeans(clazz, new AnnotationLiteral<Any>() {});
+                if (participantBeans.isEmpty()) {
+                    // resource is not registered as managed bean so register a custom managed instance
+                    try {
+                        participant.setInstance(clazz.newInstance());
+                    } catch (InstantiationException | IllegalAccessException e) {
+                        LRALogger.i18NLogger.error_cannotProcessParticipant(e);
+                    }
+                }
+            }
+        }
+
+        event.addBean()
+            .read(beanManager.createAnnotatedType(LRAParticipantRegistry.class))
+            .beanClass(LRAParticipantRegistry.class)
+            .scope(ApplicationScoped.class)
+            .createWith(context -> new LRAParticipantRegistry(participants));
+    }
+
+    /**
+     * Collects all non-JAX-RS participant methods in the defined Java class
+     *
+     * @param javaClass a class to be scanned
+     * @return Collected methods wrapped in {@link LRAParticipant} class or null if no non-JAX-RS methods have been found
+     */
+    private LRAParticipant processJavaClass(Class<?> javaClass) {
+        if (isNotLRAParticipant(javaClass)) {
+            return null;
+        }
+
+        LRAParticipant participant = new LRAParticipant(javaClass);
+
+        Arrays.stream(javaClass.getDeclaredMethods()).forEach(m -> processParticipantMethod(m, participant));
+
+        boolean shouldProcess = participant.getCompensateMethod() != null ||
+            participant.getCompleteMethod() != null ||
+            participant.getStatusMethod() != null ||
+            participant.getForgetMethod() != null;
+
+        return shouldProcess ? participant : null;
+    }
+
+    private boolean isNotLRAParticipant(Class<?> javaClass) {
```

Comment:
I found this method difficult to parse. Why not just:

```java
    private boolean isNotLRAParticipant(Class<?> javaClass) {
        for (Method method : javaClass.getMethods())
            if (method.isAnnotationPresent(LRA.class) || method.isAnnotationPresent(Compensate.class))
                return true;
        }
       return false;
   }
```

Reply:
but we need both `@LRA` and `@Compensate` method in class for it to be LRA participant if I'm not mistaken. I'm a little confused by your suggestion. For instance:

```java
@Path("not-lra")
public class NotLRAParticipant {
  @GET
  @Path("status")
  @Status
  public void status() {
    // ...
  }
}
```

The `isNotLRAParticipant(NotLRAParticipant.class)` returns false which indicates that NotLRAParticipant is a participant which not true.

Reply:
Ah you are right, both annotations need to be present (not either). 

Why not use the jandex index and the ClassInfo objects you generated in the AfterBeanDiscovery observer since that contains all the data you require and has query methods. I hope you didn't mind but I took a look at get the same info from the index:

```java
public class CDIExtension {
    private Index index;
    ...
    void processDeployment() {
        for (ClassInfo ci : index.getKnownClasses()) {
            LRAParticipant p = getAsParticipant(ci);

            if (p != null && p.shouldProcess()) {
                // update participants and add as a CDI bean
                System.out.println(p);
            }
        }
    }

    private LRAParticipant getAsParticipant(ClassInfo ci) {
        // isNotLRAParticipant
        if (!hasAnnotations(ci, Compensate.class.getName(), LRA.class.getName()))
            return null;

        // isJaxRsMethod
        if (hasAnnotations(ci,
                GET.class.getName(),
                POST.class.getName(),
                PUT.class.getName(),
                DELETE.class.getName(),
                HEAD.class.getName(),
                OPTIONS.class.getName())) {
            return null;
        }

        LRAParticipant participant = new LRAParticipant(ci.getClass());

        // setParticipantAnnotation
        participant.setCompensateMethod(ci.annotations().get(
                  DotName.createSimple(Compensate.class.getName())));
        participant.setCompleteMethod(ci.annotations().get(
                  DotName.createSimple(Complete.class.getName())));
        participant.setStatusMethod(ci.annotations().get(
                  DotName.createSimple(Status.class.getName())));
        participant.setForgetMethod(ci.annotations().get(
                  DotName.createSimple(Forget.class.getName())));

        return participant;
    }

    private boolean hasAnnotations(ClassInfo ci, String... annotations) {
        for (String annotation : annotations) {
            if (!hasAnnotation(ci, DotName.createSimple(annotation)))
                return false;
        }

        return true;
    }

    private boolean hasAnnotation(ClassInfo ci, DotName name) {
        return ci.annotations().containsKey(name);
    }
}
```

Reply:
And instead of having participant setter methods just do

>         LRAParticipant participant = new LRAParticipant(ci);

and set the methods internally in the constructor.

Reply:
My suggested getAsParticipant() method is checking for JAX-RS annotations on the whole class instead of on each method. To correct that and to check each method do something like:

```java
    private LRAParticipant getAsParticipant(ClassInfo ci) throws ClassNotFoundException {
        // isNotLRAParticipant
        if (!hasAnnotations(ci, Compensate.class.getName(), LRA.class.getName()))
            return null;

        LRAParticipant participant = new LRAParticipant(ci);
        for (MethodInfo mi : ci.methods()) {
            // isJaxRsMethod
            if (!hasAnyAnnotation(mi,
                    GET.class.getName(),
                    POST.class.getName(),
                    PUT.class.getName(),
                    DELETE.class.getName(),
                    HEAD.class.getName(),
                    OPTIONS.class.getName())) {
                // process the class if has any participant annotation:
                participant.process(mi);
            }
        }

        return participant;
    }
```
You can still use your existing method validation code by doing:

```java
        id = Class.forName(ci.name().toString());
```
to get the java class instead of the jandex ClassInfo. Alternatively the jandex type MethodInfo should give you sufficient information to do the validation (but I did not try it).

Reply:
I will try to rewrite it to use only Jandex, but as you mentioned, we will still need to cycle through declared methods I am afraid. Also, we will need to keep the Method reference, which I guess must by somehow obtainable from MethodInfo. Well, I will try and let you know if it's possible to rewrite to Jandex. But the current solution can be reused even in case there is no Jandex which may come in handy when reintroduce programmatic registration again. Otherwise, we will have to create the Jandex index dynamically, which I am not sure may or may not be a problem.

Reply:
@mmusgrov I finally got to try to rewrite annotation scanning to Jandex and I run into a fundamental problem that there is no way how to obtain reflection information from Jandex or the other way around. I actually spoke even to Jandex creator and he confirmed that Jandex is designed to run before the reflection kicks in. I was actually able to get it running but its rather complex and I must somehow "guess" the methods in participant class because there is no way how to obtain Method object from MethodInfo (switch on the number of arguments as now all parameters must be Strings so method is compensate(), compensate(String) or compensate(String, String) -- which is not really maintainable in a long term). Another reason is that we would need to dynamically create Jandex Type objects for every parameter and return type in every method instead of simple equals() on class objects. However, it does help the check whether a class is a participant (handy annotations() method you've suggested) but I don't think it's worth creating index just for this. I can update the PR, that if the index present we can optimize this part if you want. Or I can finish up the rewrite to Jandex if you would like to compare.

After all, I think the solution like it stands now is the safest bet.

Reply:
> After all, I think the solution like it stands now is the safest bet.

The jandex code that I proposed does most of what you need. The final step of getting the method info to do the invocations can be done using your existing code as I suggested in my comment https://github.com/jbosstm/narayana/pull/1450#discussion_r301944907

Reply:
My motivation for this conversation was the initial review comment

> I found this method difficult to parse ...

and when I noticed that you had already created a Jandex index I gave an example of how to simplify the method isNotLRAParticipant() and the other methods in this class using the contents of the index. This had the benefit of:

- making the code more readable
- more efficient since you already had the index

In the comment #1450 of this conversation I said "You can still use your existing method validation code (that's in LRAParticipant) by doing:  id = Class.forName(ci.name().toString()); to get the java class instead of the jandex ClassInfo.

To summarise, since you have already gone to the trouble of creating an index I would use that for all annotation processing. And for refection (since as you state is not available from the index) I would use Class.forName(ci.name().toString()) and the code you wrote for LRAParticipant.

Reply:
Note that the code I included in my previous comment is using the Jandex index for the annotation processing so we know it is feasible.